### PR TITLE
Produce withdrawals for fully matched orders

### PR DESCRIPTION
**Pull request by dat-incognito-org**

### CODE REVIEW

File: blockchain/pdex/state_v2.go [hunk 1]
```diff
@@ -338,6 +338,13 @@ func (s *stateV2) BuildInstructions(env StateEnvironment) ([][]string, error) {
 	}
 	instructions = append(instructions, tradeInstructions...)
 
+	var matchedWithdrawInstructions [][]string
+	matchedWithdrawInstructions, s.poolPairs, err = s.producer.withdrawAllMatchedOrders(s.poolPairs)
```
Comment (tin-incognito) [inline blockchain/pdex/state_v2.go:342]: This `s.poolPairs` variable will force us to run through all pool pairs, although this pool doesn't have any orders or crossed orders? Can we store which pool pair have crossed orders?

File: blockchain/pdex/state_producer_v2.go [hunk 2]
```diff
@@ -666,6 +691,57 @@ TransactionLoop:
...
+			matchedToken1 := ord.TradeDirection() == byte(v2utils.TradeDirectionSell1) &&
+				ord.Token1Balance() == 0 && ord.Token0Balance() > 0
+			if matchedToken0 {
+				currentBalance := ord.Token1Balance()
+				ord.SetToken0Balance(0)
+				withdrawResults[pair.state.Token1ID()] = currentBalance
+			}
+			if matchedToken1 {
+				currentBalance := ord.Token0Balance()
+				ord.SetToken1Balance(0)
```
Comment (cuong-incognito) [inline blockchain/pdex/state_producer_v2.go:712]: Same as above

File: blockchain/pdex/state_producer_v2.go [hunk 3]
```diff
@@ -666,6 +691,57 @@ TransactionLoop:
...
+	for pairID, pair := range pairs {
+		for _, ord := range pair.orderbook.orders {
+			withdrawResults := make(map[common.Hash]uint64)
+			matchedToken0 := ord.TradeDirection() == byte(v2utils.TradeDirectionSell0) &&
+				ord.Token0Balance() == 0 && ord.Token1Balance() > 0
+			matchedToken1 := ord.TradeDirection() == byte(v2utils.TradeDirectionSell1) &&
+				ord.Token1Balance() == 0 && ord.Token0Balance() > 0
+			if matchedToken0 {
+				currentBalance := ord.Token1Balance()
+				ord.SetToken0Balance(0)
```
Comment (cuong-incognito) [inline blockchain/pdex/state_producer_v2.go:707]: Why do we have to set token0 balance to zero when it is already zero?

File: blockchain/pdex/state_producer_v2.go [hunk 4]
```diff
@@ -666,6 +691,57 @@ TransactionLoop:
 	return result, pairs, nil
 }
 
+func (sp *stateProducerV2) withdrawAllMatchedOrders(
```
Comment (cuong-incognito) [inline blockchain/pdex/state_producer_v2.go:694]: We have yet to limit the number of auto order withdrawal instructions for each block, haven't we?

File: blockchain/pdex/state_producer_v2.go (hunk 3 above)
Comment (cuong-incognito) [reply blockchain/pdex/state_producer_v2.go:707]: Did you mean `ord.SetToken1Balance(0)`?

File: blockchain/pdex/state_producer_v2.go (hunk 4 above)
Comment (tin-incognito) [reply blockchain/pdex/state_producer_v2.go:694]: Yes it will be in other PR

File: blockchain/pdex/state_producer_v2.go (hunk 3 above)
Comment (dat-incognito-org) [reply blockchain/pdex/state_producer_v2.go:707]: updated in later commit

File: blockchain/pdex/state_v2.go (hunk 1 above)
Comment (dat-incognito-org) [reply blockchain/pdex/state_v2.go:342]: can't imagine orderbooks getting too big to traverse fast. This is simpler logic & means trading doesn't have to perform any differently. Said optimization can be done later on if needed.

**Review by tin-incognito — APPROVED** [review]
Ok

File: blockchain/pdex/state_producer_v2.go (hunk 3 above)
Comment (dat-incognito-org) [reply blockchain/pdex/state_producer_v2.go]: Nvm, thanks for spotting this. Refer to the newer commit.

**Review by cuong-incognito — APPROVED** [review]

**Pull request merged**